[pytorch] Allow async work in rpc RequestCallback processing.

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30637 [pytorch] Allow async work in rpc RequestCallback processing.**

RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

Commit:
[pytorch] Proposal: allow async work in rpc RequestCallback processing.

RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Proposal: allow async work in rpc RequestCallback processing."


RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Proposal: allow async work in rpc RequestCallback processing."


RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Proposal: allow async work in rpc RequestCallback processing."


RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

jjlilley commented:
This change should be ready for eyes:
 - I think we need a mechanism like this, so that we don't block indefinitely in hundreds of threads, especially in distributed autograd.
-  If there's an alternate approach that's better, my feelings won't be hurt :)
  But rather, I thought I'd write up a concrete PR for an approach, to get the ball rolling. Since I think we're going to continue to get bitten by issues related to zillions of threads if we don't address this.

rohan-varma commented on torch/csrc/distributed/rpc/request_callback.cpp:
```diff
@@ -28,16 +28,24 @@ struct ClearAutogradContextGuard {
-    return processMessage(request);
+    // do we need to run clear() in the future invocation?
+    auto ret = processMessage(request);
+    if (!ret->completed()) {
```
If I understand this PR correctly, `ret->completed()` should always be true, right? Since when we wrap a message into a Future, we set the completed flag in the constructor. Is there an example of how this change can be used to accommodate a request that should be completed asynchronously?

jjlilley replied:
You're right, so I went ahead and added support for BACKWARD_AUTOGRAD_REQ in this change.

It seemed to be mostly a matter of making DistEngine::executeSendFunction() and the underlying clearAndWaitForOutstandingRpcs() return a future.

Commit:
Update on "[pytorch] Proposal: allow async work in rpc RequestCallback processing."


RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/future_message.cpp:
```diff
@@ -4,6 +4,11 @@ namespace torch {
 namespace distributed {
 namespace rpc {
 
+FutureMessage::FutureMessage(Message message) {
```
Thanks, changed this.

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -370,7 +370,15 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+          } else {
+            auto fromId = work.from_.id_;
+            futureResponse->addCallback([this, fromId](const Message& msg) {
+              send(getWorkerInfo(fromId), std::move(const_cast<Message&>(msg)));
```
That's the issue with the FutureMessage api as-is - we get(const Message&), though we know that we're the only (or last) one using that FutureMessage, since we don't expose it externally.

My feeling is that changing to non-const in the api is probably worse here, since we allow a list of callbacks.

jjlilley replied:
ah, nevermind - I just updated to remove the const_cast<>.

(I was originally concerned with ref-cycle loops if I placed a reference to the future in the callback captures, but given that they're cleared afterwards, it's not an issue.)

pietern replied:
Thanks,  @jjlilley!

pritamdamania87 reviewed (commented):
Thanks for looking into this! I have one concern around the autograd context in this PR.

jjlilley commented on torch/csrc/distributed/rpc/request_callback.cpp:
```diff
@@ -28,16 +28,24 @@ struct ClearAutogradContextGuard {
+    auto ret = processMessage(request);
+    if (!ret->completed()) {
+      // If not yet complete, clear the context when we do.
+      ret->addCallback(
```
We aren't using thread-local for context, are we?
 - I did just tweak this to avoid clearing context inline in this !completed() case.
 - I don't necessarily expect that we will necessarily implement Python UDFs server-side in a pure async manner - though maybe this is reasonable to do.
 - We now won't try to clear until after the future message comes back... but not sure if this is sufficient

jjlilley commented on torch/csrc/distributed/rpc/future_message.cpp:
```diff
@@ -16,14 +21,27 @@ const Message& FutureMessage::wait() {
   return message_;
 }
 
+const Message& FutureMessage::waitNoThrow() {
```
Ah, this was needed at one point, but no longer, thanks.

Commit:
Update on "[pytorch] Proposal: allow async work in rpc RequestCallback processing."


RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Proposal: allow async work in rpc RequestCallback processing."


RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

jjlilley commented:
I need to take a further look at this; tests started failing with the revisions.

pietern commented on torch/csrc/distributed/autograd/context/context.cpp:
```diff
@@ -119,6 +119,32 @@ void DistAutogradContext::clearAndWaitForOutstandingRpcs() {
+  if (outStandingRpcs.empty()) {
+    state->future->markCompleted();
+  }
+  return state->future;
```
Nice pattern, keeping the state struct local to the function and closing on a shared_ptr.

pietern commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -222,14 +223,20 @@ void DistEngine::executeSendFunction(
-
-    // Wait for all of the outstanding rpcs to complete.
-    autogradContext->clearAndWaitForOutstandingRpcs();
+    guard.cancel();
```
This guard can be removed altogether from the looks of it.

jjlilley replied:
Yes, I agree.

pietern commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -222,14 +223,20 @@ void DistEngine::executeSendFunction(
+    // Return future waiting for all of the outstanding rpcs to complete.
+    auto future = autogradContext->clearAndWaitForOutstandingRpcsAsync();
+    future->addCallback([contextId](const rpc::Message&) {
+      ClearContextIdGuard guard2(contextId);
```
I would turn this into an explicit call to clear the context ID. The RAII wrapper does the same on destruction, which in this case is one level of indirection too many in my opinion. If it's too much code or duplication, we could consider adding a helper to clear the context ID of the current distributed autograd instance. cc @pritamdamania87

jjlilley replied:
Right, I agree. Actually, I rendered this section back to the original.

At this point, my thinking is, the right thing is to clear for the current calling thread (as originally), and if the handler is in fact async, for that specific handler to transfer the thread_local context to the async thread. If there's a generalizable pattern with respect to this, we should generalize it when it becomes more clear.

pietern commented on torch/csrc/distributed/autograd/engine/dist_engine.h:
```diff
@@ -109,11 +109,18 @@ class ClearContextIdGuard {
 
  private:
   int64_t contextId_;
+  bool valid_{true};
```
If you remove the `cancel()` call above, this can be removed as well.

pietern commented on torch/csrc/distributed/rpc/future_message.h:
```diff
@@ -12,20 +12,21 @@ namespace rpc {
 struct TORCH_API FutureMessage final {
  public:
   using Callback = std::function<void(const Message&)>;
+  FutureMessage() = default;
```
Not a fan of default constructors. What's the motivation to add it?

jjlilley replied:
The default constructor was already there previously, but implicitly defined.
Adding the explicit single-argument constructor in this change seems to require explicitly declaring that an non-arg constructor (for an unrealized future) is allowed as well.

pietern commented on torch/csrc/distributed/rpc/request_callback.cpp:
```diff
@@ -17,27 +17,38 @@ namespace {
   try {
-    return processMessage(request);
+    // do we need to run clear() in the future invocation?
+    ClearAutogradContextGuard guard;
```
Same comments regarding guards and cancelling hold here.

jjlilley replied:
Right, reverted this to the original.

Commit:
Update on "[pytorch] Allow async work in rpc RequestCallback processing."


RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Allow async work in rpc RequestCallback processing."


RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Allow async work in rpc RequestCallback processing."


RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

jjlilley commented:
From the discussion earlier today, it seems sensible to separate the infra/future work in this change, from the dist_autograd part of the work.

Hence, I've removed a few active parts of the change, so that it's mostly an infra change, but the unittests should be green (and lower risk).

It should be ready for a look.

Meanwhile, I'll clean up the separate dist_autograd component. The thread_local aspect is complicating things a bit, and doesn't seem to belong in the same change as the rest of the plumbing code here.

Thanks!

zhaojuanmao commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -224,7 +224,10 @@ void DistEngine::executeSendFunction(
+    // TODO: we currently implicitly rely this stack frame not going away
+    // before the rpcs complete (e.g. if we return the future and call
+    // wait from the caller, tests will fail).
+    autogradContext->clearAndWaitForOutstandingRpcsAsync()->waitNoThrow();
```
@jjlilley I'm wondering why it throws exception here? if it is expected, is it better to catch exception here and log? in this case, we do not need waitNoThrow() API?

jjlilley replied:
I'll get rid of it here, since there's actually no way that it will throw as-is.
The change is still using one instance of waitNoThrow() in request_callback_impl, though once we get rid of it (future change), I can GC the waitNoThrow() api :)

zhaojuanmao commented on torch/csrc/distributed/rpc/future_message.cpp:
```diff
@@ -16,14 +19,28 @@ const Message& FutureMessage::wait() {
+    lock.unlock();
+    for (auto& callback : cbs) {
+      callback(message_);
+    }
```
callbacks_.clear()?

jjlilley replied:
callbacks_.clear() is implicit here because of the std::move() from the vector.
That said, I'll replace with a swap with an empty vector, which should be more explicit.

zhaojuanmao commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -375,7 +375,18 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+          } else {
+            auto fromId = work.from_.id_;
+            futureResponse->addCallback(
+                [this, fromId, futureResponse](const Message&) {
```
nit: no need "this"?

jjlilley replied:
We need this for the send() call.

Commit:
Update on "[pytorch] Allow async work in rpc RequestCallback processing."


RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -29,9 +29,15 @@ namespace rpc {
+    int64_t messageId) const {
+  auto wrap = [messageId](Message m) {
+    m.setId(messageId);
+    return std::make_shared<FutureMessage>(std::move(m));
```
sorry, I'm a little confused about how this is async. Would you please explain more? IIUC, when the futureMessage is created here, message has already been processed and is completed, this looks like it is the same as current sync behavior. I thought we should immediately return a futureMessage in some way and mark the message later on/asynchronously?

jjlilley replied:
Right, the parts exiting with wrap() and returning a completed FutureMessage are definitely not async :)

In a previous iteration, I had stacked in dist autograd changes which did return an async FutureMessage, though seemed sensible to separate out to reduce merge risk.  (Still trying to gauge this team's culture with respect to spanning changes)

jjlilley replied:
Actually, right after typing this, I just got tests to pass with my async BACKWARD_AUTOGRAD_REQ impl (I had returned an incomplete future in a base case that I should have returned a completed one).

So, I updated this change to include async for BACKWARD_AUTOGRAD_REQ, which should be a win here,

Commit:
Update on "[pytorch] Allow async work in rpc RequestCallback processing."


RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

jjlilley commented:
Any other thoughts here? Thanks.

pietern commented:
@jjlilley Can you take a look at my comments above? Feel free to hit resolve when addressed.

jjlilley reviewed (commented):
@pietern Sorry, I wasn't aware of this aspect of review etiquette here! I've been used to working largely in Phabricator that last several years...

I believe I've addressed the comments above, added a bit of clarification, resolved threads. Please let me know if there's anything else needed from me.

mrshenli commented on torch/csrc/distributed/rpc/future_message.cpp:
```diff
@@ -16,14 +19,29 @@ const Message& FutureMessage::wait() {
-    fireCallbacks();
+    completed_ = true;
+    std::vector<Callback> cbs;
+    cbs.swap(callbacks_);
```
nice!

mrshenli commented on torch/csrc/distributed/rpc/future_message.cpp:
```diff
@@ -16,14 +19,29 @@ const Message& FutureMessage::wait() {
   return message_;
 }
 
+const Message& FutureMessage::waitNoThrow() {
```
Question, why we would want to ignore the exception message for `FORWARD_AUTOGRAD_REQ`?

jjlilley replied:
The intention here isn't to ignore the exception, but rather to pass it through, and allow it to be processed at a higher layer. If this isn't appropriate, I'm happy to change, but was concerned about over-processing at this layer.

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -375,7 +375,18 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+          if (futureResponse->completed()) {
+            send(work.from_, std::move(*futureResponse).moveMessage());
+          } else {
+            auto fromId = work.from_.id_;
```
Would I be correct if I assume current this path is not covered by any test/code?

jjlilley replied:
It's covered by caffe2/test:dist_autograd_spawn  (if I add add a LOG(FATAL), it's reached there)

I implemented the one async autograd op partly to get test coverage..

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -53,16 +59,17 @@ Message RequestCallbackImpl::processRpc(
           stack.size());
 
-      return std::move(ScriptResp(std::move(stack.front()))).toMessage();
+      return wrap(std::move(ScriptResp(std::move(stack.front()))).toMessage());
```
I wonder if it would be more laconic if we do wrap in `RequestCallbackImpl::processMessage` instead of in every switch-case here?

jjlilley replied:
- The problem is that we can only set the messageId() (as wrap does) once the future is completed. Creating an additional future hop for this feels suboptimal. This was the least verbose way I thought of doing for the non-async cases.

- I don't think I remember using "laconic" since I was 17 and studying for the SATs :)

mrshenli commented on torch/csrc/distributed/autograd/context/context.cpp:
```diff
@@ -108,15 +108,30 @@ void DistAutogradContext::addOutstandingRpc(
+  auto state = std::make_shared<State>(outStandingRpcs.size());
+  for (auto& rpc : outStandingRpcs) {
+    rpc->addCallback([state](const rpc::Message&) {
+      if (--state->remaining == 0) {
```
callbacks can fire concurrently, do we need a lock here?

jjlilley replied:
no, this is an atomic, just need to decrement and process when one hits zero.

mrshenli commented on torch/csrc/distributed/autograd/context/context.cpp:
```diff
@@ -108,15 +108,30 @@ void DistAutogradContext::addOutstandingRpc(
+      }
+    });
   }
+  if (outStandingRpcs.empty()) {
```
Why the above for loop is not in the else clause of this if clause?

jjlilley replied:
I can change as such... I was just trying to avoid excess indentation

mrshenli reviewed (commented):
Thanks a lot for working on this! This looks great! I just have some questions inline.

Commit:
rebase and minor context.cpp if change on "[pytorch] Allow async work in rpc RequestCallback processing."


RequestCallback api currently forces work to be always synchronous, which,
as we scale, means we're going to need to throw large number of (mostly
blocked) threads at the rpc problem. For some activities like dependent
autograd rpcs, there's not a necessary reason to block in these threads.

In this change, the RequestCallback api is updated to return a
shared_ptr<FutureMessage> rather than a Message:

   std::shared_ptr<FutureMessage> operator()(Message& request) const;

With a futures-style api, RPC ops that wish to be async can then be async,
while short-lived blocking functions (or Python UDFs) can just block.

In this change, we keep all of the current ops as synchronous (i.e. we block
and then return a completed FutureMessage). We also update the rpc_agents in
a manner compatible with this sort of parallelism.

Here, we only want to incur overhead when we use the async behavior.
Some modest extra cost seems unavoidable here (e.g. the allocation for the
std::make_shared<>), but we can trivially detect the synchronous/completed
case in the rpc_agent and avoid the extra thread-switches/etc. in that case.

Differential Revision: [D18774322](https://our.internmc.facebook.com/intern/diff/D18774322/)

[ghstack-poisoned]

jjlilley commented:
Thanks for the review comments, Shen. Uploaded a rebased version with the context.cpp tweak.

mrshenli reviewed (approved):
Thank you!

zhaojuanmao commented:
@jjlilley I will rebase my Future diffs after you land

jjlilley commented:
> @jjlilley I will rebase my Future diffs after you land

Thanks, @zhaojuanmao - trying to land quite shortly this afternoon.

jjlilley commented:
> Thank you!

Thanks for the review!

pietern commented:
@jjlilley Thanks for resolving! I've gotten used to the comments being resolved meaning they're acknowledged (not necessarily resulting in a change). Not seeing them resolved makes me wonder if they were seen at all.
